Feature/update joysticks wiki (#28)
OpenSpace/openspace.github.io: docs/builders/joystick-customization.md

---
title: Joystick Customization
layout: default

parent: Builders
nav_order: 4
---

# Customizing the joystick navigation

> This page will go in-depth into how to customize the joystick navigation to your own liking. If you want to learn more about the default joystick navigation in OpenSpace, see [Joystick Navigation](../../users/navigation/joysticks).

Reviewer: ```suggestion
This page goes in-depth into how to customize the joystick navigation to your own liking. If you want to learn more about the default joystick navigation in OpenSpace, see [Joystick Navigation](../../users/navigation/joysticks).
```

| "Zoom"       | Move the camera closer or further away from the focus                   |
| "Zoom In"    | Move the camera closer to the focus                                     |
| "Zoom Out"   | Move the camera further away from the focus                             |
| "LocalRoll"  | Roll the camera (clockwise) in relation to the middle of the view       |
| "GlobalRoll" | Roll the camera (clockwise) in relation to the focus                    |
| "Pan X"      | Turn the camera left/right in relation to the view                      |
| "Pan Y"      | Turn the camera up/down in relation to the view                         |

4. (Optional) Whether or not this axis should be inverted. This is a common setting in video games. Defaults to `false`.

5. (Optional) The type of joystick that this axis represents on the controller. The options are either `"JoystickLike"` or `"TriggerLike"`. A joystick is `"TriggerLike"` if it can only be pressed or pushed in one direction. A `"JoystickLike"` axis can be pushed in two directions, for example, left **and** right, or up **and** down. Defaults to `"JoystickLike"`.

> 6. (Optional) Whether or not this axis is "sticky". In most cases, this should be set to `false`. An axis is "sticky" if, when you let go of it, the values it represents in the software do not go back to the default. Another sign is that the longer you push it, the value, and therefore the movement, gets more and more extreme over time and does not stop or slow down when you let go. Defaults to `false`.

Reviewer: This is not super clear;  maybe making the connection to a throttle on a joystick?

Author: Which one are you referring to? Argument 5 or 6? I'm not sure a throttle is a good example for number 6 but it could be for number 5 instead, although there I think the trigger is an easier example.

Number 6 is really hard to explain in a brief way because it is all about the values that are sent and has nothing to do with how the actual joystick looks like. It can look like it goes back to the "origin" when it is let go, but the values it sends don't go back, they just stay where they were. And when you press the joysticks longer over time the values just keep going even beyond the -1 to 1 range, it could go on to really high values. So the value we use instead then is the difference between the new and the old value instead of the value itself. We sort of re-normalize it every frame instead of taking the raw value.

Reviewer: Ah.. yea, I meant option 6..   so in that case let's leave it as it is

8. (Optional) Sensitivity value for this axis. Can be used to fine-tune the sensitivity for all axes individually. Defaults to `1.0`. A value larger than `1.0` would lead to the axis becoming more sensitive and a value that is smaller than `1.0` would lead to the axis being less sensitive to input. There is also a global sensitivity property that can be adjusted in the GUI, under *Settings*, *Navigation Handler*, *Orbital Navigator* and then *Joystick Sensitivity* (identifier `NavigationHandler.OrbitalNavigator.JoystickSensitivity`). Note that this will affect all connected controllers and all of their axes.

Here is an example asset with the SpaceMouse:

> ~~~lua

Reviewer: ```suggestion
```lua
```

  local SpaceMouse = {
    -- Axes
    Push = {
      LeftRight = 0,
      BackForth = 1,
      UpDown = 2
    },
    Twist = 5,
    Tilt = {
      LeftRight = 4,
      BackForth = 3
    },

    -- Buttons
    LeftButton = 0,
    RightButton = 1
  }

  asset.onInitialize(function()
    local controller = SpaceMouse;
    local name = "SpaceNavigator";

    openspace.navigation.bindJoystickAxis(name, controller.Push.LeftRight, "Orbit X", false, "JoystickLike", true, false, 40.0)
    openspace.navigation.bindJoystickAxis(name, controller.Push.BackForth, "Orbit Y", false, "JoystickLike", true, false, 40.0)
    openspace.navigation.bindJoystickAxis(name, controller.Twist, "Pan X", true, "JoystickLike", true, false, 40.0)
    openspace.navigation.bindJoystickAxis(name, controller.Tilt.BackForth, "Pan Y", false, "JoystickLike", true, false, 35.0)
    openspace.navigation.bindJoystickAxis(name, controller.Push.UpDown, "Zoom", false, "JoystickLike", true, false, 40.0)
    openspace.navigation.bindJoystickAxis(name, controller.Tilt.LeftRight, "LocalRoll", false, "JoystickLike", true, false, 35.0)
  end)

> ~~~

Reviewer: -> ```

3. The full identifier for the property you want to control with this axis on the controller. A tip to find this identifier is to first run OpenSpace normally and change the property with the normal GUI, then you can look at the Script Log and see the full name of the property you just changed, for more information about the Script Log see [The Script Log](../../users/commandline#the-script-log).

4. (Optional) The minimum value allowed to be set for this property using the axis. Defaults to `0.0`.

5. (Optional) The maximum value allowed to be set for this property using the axis. Defaults to `100.0`.

6. (Optional) Whether or not this axis should be inverted. This is a common setting in video games. Defaults to `false`.

7. (Optional) Whether or not the property change should be forwarded to other connected nodes or sessions. This is similar to the `"isLocal"` parameter for actions. Defaults to `true`.

Here is an example asset with the Earth scale bound to the right trigger on an Xbox controller:

> ~~~lua

Reviewer: ```suggestion
```lua
```

  local XBoxController = {
    -- Axes
    LeftThumbStick = {
      LeftRight = 0,
      UpDown = 1
    },
    RightThumbStick = {
      LeftRight = 2,
      UpDown = 3
    },
    LeftTrigger = 4,
    RightTrigger = 5,

    -- Buttons
    A = 0,
    B = 1,
    X = 2,
    Y = 3,
    LB = 4,
    RB = 5,
    Back = 6,
    Start = 7,
    LeftStickButton = 8,
    RightStickButton = 9,
    DPad = {
      Up = 10,
      Right = 11,
      Down = 12,
      Left = 13
    }
  }

  asset.onInitialize(function()
    local controller = XBoxController;
    local name = "Xbox Controller";

    -- Bind Right trigger to Earth Scale
    openspace.navigation.bindJoystickAxisProperty(name, controller.RightTrigger, "Scene.Earth.Scale.Scale", 0.1, 100, false, true);
  end)

> ~~~

Reviewer: -> ```

3. The script that should be executed when the button is activated. A tip to find this script is to use the Script Log, for more information see [The Script Log](../../users/commandline#the-script-log).

4. Description of the script that the button will execute when the button is activated.

5. (Optional) When the button should be interpreted as activated, defaults to `"Press"`. This must be one of the identifiers in the following list:

| Identifier | Description                                     |
| ---------- | ----------------------------------------------- |
| "Idle"     | When the button is **not** pressed              |
| "Press"    | When the button is pressed                      |
| "Repeat"   | When the button is held pressed                 |
| "Release"  | When the button was pressed and is not released |

6. (Optional) Whether or not the script should be forwarded to other connected nodes or sessions. This is similar to the `"isLocal"` parameter for actions. Defaults to `true`.

Here is an example asset that switches focus when pressing the trigger buttons on an Xbox controller:

> ~~~lua

Reviewer: ```suggestion
```lua
```

  local XBoxController = {
    -- Axes
    LeftThumbStick = {
      LeftRight = 0,
      UpDown = 1
    },
    RightThumbStick = {
      LeftRight = 2,
      UpDown = 3
    },
    LeftTrigger = 4,
    RightTrigger = 5,

    -- Buttons
    A = 0,
    B = 1,
    X = 2,
    Y = 3,
    LB = 4,
    RB = 5,
    Back = 6,
    Start = 7,
    LeftStickButton = 8,
    RightStickButton = 9,
    DPad = {
      Up = 10,
      Right = 11,
      Down = 12,
      Left = 13
    }
  }

  asset.onInitialize(function()
    local controller = XBoxController;
    local name = "Xbox Controller";

    -- Switch focus in the interesting nodes list
    -- Previous
    openspace.navigation.bindJoystickButton(
      name,
      controller.LB,
      "openspace.navigation.targetPreviousInterestingAnchor(); openspace.navigation.retargetAnchor();",
      "Switch target to the previous interesting node",
      "Press",
      true
    )

    -- Next
    openspace.navigation.bindJoystickButton(
      name,
      controller.RB,
      "openspace.navigation.targetNextInterestingAnchor(); openspace.navigation.retargetAnchor();",
      "Switch target to the next interesting node",
      "Press",
      true
    )
  end)

> ~~~

Reviewer: -> ```

![](images/joystick-axis.png)

For the buttons, we will use the same strategy, push one button a few times and see which of the buttons in OpenSpace reacts, note down that number, and move on to the next. Now that we have a clear image of how each button and axis relate to the list in OpenSpace we could build a map in our asset. Below is an image to show how it can look like when one of the buttons is pressed:

![](images/joystick-button.png)

In each joystick asset file that OpenSpace provides there will be a section at the top that looks similar to this:

> ~~~lua

Reviewer: ```suggestion
```lua
```

-- Name of the controller
local XBoxController = {
  -- Axes
  LeftThumbStick = {
    LeftRight = 0,
    UpDown = 1
  },
  RightThumbStick = {
    LeftRight = 2,
    UpDown = 3
  },
  LeftTrigger = 4,
  RightTrigger = 5,

  -- Buttons
  A = 0,
  B = 1,
  X = 2,
  Y = 3,
  LB = 4,
  RB = 5,
  Back = 6,
  Start = 7,
  LeftStickButton = 8,
  RightStickButton = 9,
  DPad = {
    Up = 10,
    Right = 11,
    Down = 12,
    Left = 13
  }
}

> ~~~

Reviewer: -> ```

This is the map of the controller that tells OpenSpace how each axis and button relate to the indices in the list that OpenSpace sees. Now with all the information you now have about your controller, you can make your own map. Start with a name for your controller, in the example above it is called `XBoxController`. Then you can start with the axes, in the example above there are four joysticks that each have one to two axes. There are the two joysticks that can move both left/right, and up/down, then there are the two triggers that only have one axis each. To add an axis to the list you add its name (no spaces or special characters) and set that equal to the index it corresponded to in the OpenSpace list. So it would look like:

> ~~~lua

Reviewer: ```suggestion
```lua
```

local NameOfController = {
  -- Axes
  NameOfJoystick1 = { -- This joystick on the controller has two axes, one for left/right and one for up/down
    LeftRight = 0,
    UpDown = 1
  },
  NameOfTrigger1 = 2, -- This axis is a trigger
  ...

  -- Buttons
  NameOfButton1 = 0,
  NameOfButtons2 = 1,
  ...

  -- Collection of buttons
  NameOfButtonCollection = {
    NameOfButton1InCollection = 5,
    NameOfButton2InCollection = 6,
    ...
  }

  ...
}

> ~~~

Reviewer: -> ```
